tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+  Result<std::string> ToUrl(std::string_view path) const override {
+    return GetS3Url(data_.bucket, path);
+  }
+};
+
+class S3KeyValueStore
+    : public internal_kvstore::RegisteredDriver<S3KeyValueStore,
+                                                S3KeyValueStoreSpec> {
+ public:
+  const std::string & endpoint() const { return endpoint_; }
+  bool IsAwsEndpoint() const { return absl::EndsWith(endpoint_, kDotAmazonAwsDotCom); }
+
+  Future<ReadResult> Read(Key key, ReadOptions options) override;
+
+  Future<TimestampedStorageGeneration> Write(Key key,
+                                             std::optional<Value> value,
+                                             WriteOptions options) override;
+
+  void ListImpl(ListOptions options,
+                AnyFlowReceiver<absl::Status, Key> receiver) override;
+
+  Future<const void> DeleteRange(KeyRange range) override;
+
+  absl::Status GetBoundSpecData(SpecData& spec) const {
+    spec = spec_;
+    return absl::OkStatus();
+  }
+
+  const Executor& executor() const {
+    return spec_.data_copy_concurrency->executor;
+  }
+
+  RateLimiter& read_rate_limiter() {
+    if (spec_.rate_limiter.has_value()) {
+      return *(spec_.rate_limiter.value()->read_limiter);
+    }
+    return no_rate_limiter_;
+  }
+  RateLimiter& write_rate_limiter() {
+    if (spec_.rate_limiter.has_value()) {
+      return *(spec_.rate_limiter.value()->write_limiter);
+    }
+    return no_rate_limiter_;
+  }
+
+  RateLimiter& admission_queue() { return *spec_.request_concurrency->queue; }
+
+  Result<std::optional<S3Credentials>> GetCredentials() {
+    if(!credential_provider_) {
+      auto result = GetS3CredentialProvider(spec_.profile, transport_);
+      if(!result.ok() && absl::IsNotFound(result.status())) {
+        credential_provider_ = nullptr;
+      } else {
+        TENSORSTORE_RETURN_IF_ERROR(result);
+        credential_provider_ = std::move(*result);
+      }
+    }
+    if(!*credential_provider_) return std::nullopt;
+    auto credential_result_ = (*credential_provider_)->GetCredentials();
+    if(!credential_result_.ok() && absl::IsNotFound(credential_result_.status())) {
+      return std::nullopt;
+    }
+    return credential_result_;
+  }
+
+// Apply default backoff/retry logic to the task.
+  // Returns whether the task will be retried. On false, max retries have
+  // been met or exceeded.  On true, `task->Retry()` will be scheduled to run
+  // after a suitable backoff period.
+  template <typename Task>
+  absl::Status BackoffForAttemptAsync(
+      absl::Status status, int attempt, Task* task,
+      SourceLocation loc = ::tensorstore::SourceLocation::current()) {
+    if (attempt >= spec_.retries->max_retries) {
+      return MaybeAnnotateStatus(
+          status,
+          tensorstore::StrCat("All ", attempt, " retry attempts failed"),
+          absl::StatusCode::kAborted, loc);
+    }
+
+    // https://cloud.google.com/storage/docs/retry-strategy#exponential-backoff
```

Comment:
Is there equivalent amazon documentation for backoff parameters?

Reply:
There's some exp. backoff logic that's described here:

- https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
- https://aws.amazon.com/builders-library/timeouts-retries-and-backoff-with-jitter/
- s3 does issue 503's (slow down).

I guess the gcs rate limiting functionality should probably be replaced with per-s3 endpoint rate-limiting behaviour. Have marked this in the TODO.